File: src/main/java/nz/co/searchwellington/model/frontend/FrontendNewsitem.java
package nz.co.searchwellington.model.frontend;

import java.util.Date;
import java.util.List;

import nz.co.searchwellington.model.Comment;
import nz.co.searchwellington.model.Twit;
import uk.co.eelpieconsulting.common.views.rss.RssFeedable;

public class FrontendNewsitem extends FrontendResource implements RssFeedable {
	
	private static final long serialVersionUID = 1L;
	
	private String publisherName;
	private List<Twit> retweets;
	private String acceptedFromFeedName;
	private String acceptedByProfilename;
	private List<Comment> comments;
	private Date accepted;
	private FrontendImage image;
	private List<FrontendTweet> twitterMentions;
	
	public String getPublisherName() {
		return publisherName;
	}
	
	public void setPublisherName(String publisherName) {
		this.publisherName = publisherName;
	}

	public List<Twit> getRetweets() {
		return retweets;
	}

	public void setRetweets(List<Twit> retweets) {
		this.retweets = retweets;
	}

	public String getAcceptedFromFeedName() {
		return acceptedFromFeedName;
	}

	public void setAcceptedFromFeedName(String acceptedFromFeedName) {
		this.acceptedFromFeedName = acceptedFromFeedName;
	}
	
	public List<Comment> getComments() {
		return comments;
	}

	public void setComments(List<Comment> comments) {
		this.comments = comments;
	}

	public String getAcceptedByProfilename() {
		return acceptedByProfilename;
	}

	public void setAcceptedByProfilename(String acceptedByProfilename) {
		this.acceptedByProfilename = acceptedByProfilename;
	}

	public Date getAccepted() {
		return accepted;
	}

	public void setAccepted(Date accepted) {
		this.accepted = accepted;
	}
	
	@Override
	public String getAuthor() {
		return publisherName;
	}
	
	public FrontendImage getFrontendImage() {	// TODO rename to getImage
		return image;
	}

	public void setFrontendImage(FrontendImage image) {
		this.image = image;
	}
	
	@Override
	public String getImageUrl() {
		return image != null ? image.getUrl() : null;
	}
	
	public List<FrontendTweet> getTwitterMentions() {
		return twitterMentions;
	}

	public void setTwitterMentions(List<FrontendTweet> twitterMentions) {
		this.twitterMentions = twitterMentions;
	}

	@Override
	public String toString() {
		return "FrontendNewsitem [accepted=" + accepted
				+ ", acceptedByProfilename=" + acceptedByProfilename
				+ ", acceptedFromFeedName=" + acceptedFromFeedName
				+ ", comments=" + comments + ", image=" + image
				+ ", publisherName=" + publisherName + ", retweets=" + retweets
				+ ", twitterMentions=" + twitterMentions + "]";
	}
	
}

File: src/main/scala/nz/co/searchwellington/model/frontend/FrontendNewsitem.scala
package nz.co.searchwellington.model.frontend

import com.fasterxml.jackson.annotation.JsonFormat
import nz.co.searchwellington.model.geo.Geocode
import nz.co.searchwellington.model.{Tag, User}
import uk.co.eelpieconsulting.common.views.rss.RssFeedable

import java.net.URLEncoder
import java.util
import java.util.Date
import scala.jdk.CollectionConverters._

case class FrontendNewsitem(id: String,
                            urlWords: String = null,
                            `type`: String = null,
                            name: String = null,
                            url: String = null,
                            httpStatus: Option[Int] = None,
                            date: Date = null,
                            description: String = null,
                            liveTime: Date = null,
                            tags: Option[Seq[Tag]] = Some(Seq.empty),
                            handTags: Option[Seq[Tag]] = Some(Seq.empty),
                            owner: String = null,
                            geocode: Option[Geocode] = None,
                            held: Boolean = false,
                            publisherName: Option[String] = None,
                            publisherUrlWords: Option[String] = None,
                            acceptedFrom: Option[FrontendFeed] = None,
                            acceptedBy: Option[User] = None, // TODO Frontend user
                            accepted: Date = null,
                            image: FrontendImage = null,
                            lastScanned: Option[Date] = None,
                            lastChanged: Option[Date] = None,
                            actions: Seq[Action] = Seq.empty,
                            twitterImage: String = null,
                           ) extends FrontendResource with RssFeedable {

  def getPublisherName: String = publisherName.orNull
  def getPublisherUrlWords: String = publisherUrlWords.orNull

  def getAcceptedFromFeedName: String = acceptedFrom.map(_.name).orNull

  def getAcceptedByProfilename: String = acceptedBy.flatMap(_.profilename).orNull

  @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ssXXX")
  def getAccepted: Date = accepted

  override def getAuthor: String = getPublisherName

  def getFrontendImage: FrontendImage = image

  override def getImageUrl: String = {
    if (twitterImage != null) {
      "https://cards.eelpieconsulting.co.uk/thumbnail?url=" + URLEncoder.encode(twitterImage, "UTF-8") // TODO not in a great place
    } else {
      null
    }
  }

  def getHangTags: util.List[Tag] = {
    handTags.map(_.asJava).orNull
  }

  def getTwitterImage: String = twitterImage

}
